### PR TITLE
Start working on proof of concept for exposing Backtrace in core

### PR DESCRIPTION
**Pull request by yaahc**
Implementation of the proof of concept described in https://github.com/rust-lang/project-error-handling/issues/3

### CODE REVIEW

File: library/core/src/backtrace.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    /// and slow, so these environment variables allow liberally using
+    /// `Backtrace::capture` and only incurring a slowdown when the environment
+    /// variables are set.
+    ///
+    /// To forcibly capture a backtrace regardless of environment variables, use
+    /// the `Backtrace::force_capture` function.
+    #[inline(never)] // want to make sure there's a frame here to remove
+    #[unstable(feature = "core_backtrace", issue = "74465")]
+    pub fn capture() -> Self {
+        if unsafe { backtrace_enabled() } {
```
Comment (Lokathor) [inline library/core/src/backtrace.rs:75]: Seems that this function call isn't `unsafe` in the current code version

File: library/core/src/backtrace.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    /// capture a backtrace. Capturing a backtrace can be both memory intensive
+    /// and slow, so these environment variables allow liberally using
+    /// `Backtrace::capture` and only incurring a slowdown when the environment
+    /// variables are set.
+    ///
+    /// To forcibly capture a backtrace regardless of environment variables, use
+    /// the `Backtrace::force_capture` function.
+    #[inline(never)] // want to make sure there's a frame here to remove
+    #[unstable(feature = "core_backtrace", issue = "74465")]
+    pub fn capture() -> Self {
```
Comment (KodrAus) [inline library/core/src/backtrace.rs:74]: One design that was floated in our Libs meeting was to define `Backtrace::capture` and co in `std`, so there would be no way (yet) to capture a `Backtrace` in `core`, even though we could refer to one there. That could remove the need for figuring out hooks right now. Supposedly adding impl blocks in `std` to types defined in `core` is something we can already do. An alternative option could be a free `backtrace::capture` function.

So this would look something like `RawWaker::new` where it accepts a `*mut dyn RawBacktrace`.

Comment (bjorn3) [reply library/core/src/backtrace.rs:74]: I am in favor of this idea. As added bonus it would probably (unstably) allow arbitrary backtraces to be constructed for for example language level backtraces in interpreters or jit engines. Or allow mixing of language level and native frames.

Comment (tmandry) [reply library/core/src/backtrace.rs:74]: I'm lacking context but what's left to figure out, does this code not work as written?

As I mentioned in the other thread, the compiler [has a usecase](https://github.com/rust-lang/rust/issues/79153) for capturing a backtrace in core (in signal handlers), so it would be nice if there were some unstable path we could call to do it.

Comment (KodrAus) [reply library/core/src/backtrace.rs:74]: @tmandry I think the missing piece here is a hook that makes `Backtrace::capture` do something meaningful when `std` is linked in. We could avoid that need by defining the `capture` function itself in `std` either on the `Backtrace` type or as a free function.

Does the compiler usecase need to capture a `Backtrace` type? Or just print one? We would still have an (unstable) way to build a `Backtrace` in `core`, it just wouldn't be on a path to stabilization yet.

Comment (tmandry) [reply library/core/src/backtrace.rs:74]: We just need to print one, creating an actual `Backtrace` value is not in scope. But printing is still a lot of logic I don't want to duplicate.

Comment (KodrAus) [reply library/core/src/backtrace.rs:74]: Yeh, it's a lot to keep in sync already between `panic!` and `Backtrace::capture` output too, so was wondering whether some non-allocating equivalent to the `backtrace` crate's frames iterator for printing could be worthwhile for that.

Comment (yaahc) [reply library/core/src/backtrace.rs]: I started the PR over again to try to work in the suggestions here but im currently stuck at the "implement methods on a foreign type" step. It seems we need to use a lang item to do this but @eddyb is worried that using that for anything other than a builtin type could cause issues with coherence.

Comment (bjorn3) [reply library/core/src/backtrace.rs]: It doesn't need to be a lang item. It could be a method on an extension trait defined inside libstd. `core::backtrace::Backtrace` then only has to have a public unstable constructor that can be used within the extension trait to create it.

Comment (yaahc) [reply library/core/src/backtrace.rs]: > It doesn't need to be a lang item. It could be a method on an extension trait defined inside libstd. core::backtrace::Backtrace then only has to have a public unstable constructor that can be used within the extension trait to create it.

Follow up questions:

* would users then need to bring this extension trait into scope in order to use any methods on backtrace? Would we want to add this trait to the prelude?
* Is the extension trait unstable? How do we add new methods to backtrace in the future? Is it sealed?

Comment (bjorn3) [reply library/core/src/backtrace.rs]: The trait should be stable but sealed. Whether or not to add it to the prelude I don't know.

Comment (tmandry) [reply library/core/src/backtrace.rs]: Maybe relevant: I recently discovered that `use FooTrait as _;` works inside the prelude (or some other module with a wildcard import), pulling it in scope without polluting the namespace.

Comment (yaahc) [reply library/core/src/backtrace.rs]: Question: Are we confident that people won't want to eventually trigger backtrace capture in `no_std` code bases (libraries) where the backtrace uses `std` machinery if that library is linked into an application that includes std?

I realized that the recent suggestions for ways we could accomplish the core backtrace move all seem to discard this feature, but I'm not confident that this isn't a useful feature or that going down these alternate paths might make it harder to add this feature in the future. My instinct is that I still want to lean towards the allocator lang item inspired design rather than an impl block lang item design, since either way they seem to be using the same tricks to break the coherence rules.

**Comment by yaahc** [comment]
Another thing I've been thinking about. I'm worried that a proof of concept is an insufficient step in proving that backtrace can be moved to core, and that we might also end up wasting effort or never see this project finished. As such I feel like this PR should aim to be the prototype and then eventually the final implementation that we merge. Possibly with an RFC based on the first pass implementation to approve the full design.

**Comment by yaahc** [comment]
## Update

I talked to the libs team today about this and have a plan now. More or less based on what I've said above I'm going to start with trying to aim straight for what I think the final implementation should look like, and not just for backtrace but also for the error trait in core. The plan is:

1. implement backtrace based on how `#[global_allocator]` and `#[default_lib_allocator]` work, with a new lang item
  1. where we have a global that owns the implementation of a trait for how to construct backtraces
  2. the global is declared in core and instantiated in std.
  3. the compiler will provide a default impl that always captures disabled backtraces, for when `std` is not linked in
  4. and used to implement all of the interfaces currently on `Backtrace`
2. Move Error to core, add a second lang item for its `downcast` impl, so that the downcast impl live in `std` while the error trait is defined in core.

The hope is that, given what we know now, we think this should be a relatively small amount of work. If everything goes according to plan this shouldn't delay the backtrace stabilization, and would get us `Error` in core much sooner. We are worried about possible unforeseen issues with using lang items though, so if issues pop up I will switch gears to focus on backtrace stabilization and focus on unblocking that rather than completing the whole project upfront.

File: library/core/src/backtrace.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+/// Global implementation of backtrace functinality. Called to create
+/// `RawBacktrace` trait objects.
+#[allow(improper_ctypes)]
+extern {
+    #[linkage = "extern_weak"]
+    #[link_name = "__rust_backtrace_capture"]
+    static BACKTRACE_CAPTURE: *mut ();
+
+    #[linkage = "extern_weak"]
```
Comment (bjorn3) [inline library/core/src/backtrace.rs:29]: 1. Is weak linkage supported by all object file formats that we would ever want to support? I guess not.
2. I think trying generate a backtrace when this isn't possible should give a compile time error. Otherwise you may not realize that backtraces are unavailable until you need them.

Comment (yaahc) [reply library/core/src/backtrace.rs:29]: 1. not from what I've been told
2. :+1:

I've since swapped to using a `weak_lang_item` which i think gives the property from (2)

File: library/core/src/backtrace.rs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    }
+}
+
+#[unstable(feature = "backtrace", issue = "74465")]
+impl fmt::Display for Backtrace {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let imp: &dyn RawBacktrace = unsafe { &*self.inner };
+        fmt::Display::fmt(imp, fmt)
+    }
+}
```
Comment (bjorn3) [inline library/core/src/backtrace.rs:200]: nit: missing trailing newline

File: library/core/src/backtrace.rs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    Disabled,
+    /// A backtrace has been captured and the `Backtrace` should print
+    /// reasonable information when rendered.
+    Captured,
+}
+
+// perma(?)-unstable
+#[unstable(feature = "backtrace", issue = "74465")]
+///
+pub trait RawBacktrace: fmt::Debug + fmt::Display + 'static {
```
Comment (bjorn3) [inline library/core/src/backtrace.rs:24]: Can you add a Send and Sync bound?

**Comment by lygstate** [comment]
Wonderfull work:) Waiting for it

**Comment by lygstate** [comment]
According to https://github.com/rust-lang/project-error-handling/issues/11#issuecomment-780045953

> I think `pub type Error = TinyBox<dyn Error>;` should work without alloc using #20 albeit only for a trimmed `Error` trait without backtrace.

The core:Error should also be able to using without alloc

File: library/std/src/backtrace.rs [hunk 6]
```diff
@@ -104,35 +104,103 @@ use crate::sync::Once;
 use crate::sys_common::backtrace::{lock, output_filename};
 use crate::vec::Vec;
 
+pub use core::backtrace::Backtrace;
+pub use core::backtrace::BacktraceStatus;
+use core::backtrace::RawBacktrace;
+
+/// Returns whether backtrace captures are enabled through environment
+/// variables.
+#[export_name = "__rust_backtrace_enabled"]
```
Comment (bjorn3) [inline library/std/src/backtrace.rs:113]:
```suggestion
#[export_name = "rust_backtrace_enabled"]
```
I don't think this should not use the `__` prefix. At least `rust_begin_unwind`, `rust_oom` and `rust_eh_personality` don't use it.

Comment (JDuchniewicz) [reply library/std/src/backtrace.rs:113]: Unfortunately this does not solve the compilation issues 😢

Comment (jyn514) [reply library/std/src/backtrace.rs:113]: > Unfortunately this does not solve the compilation issues cry

The compiler errors are just from the test suite, I wouldn't expect them to be super hard to fix.

Comment (yaahc) [reply library/std/src/backtrace.rs]: I ended up fixing this by using `#[lang ...]` instead of `#[export_name ...]` and now I have no idea why I even used `export_name` in the first place...

**Comment by Aaron1011** [comment]
@yaahc - Are you still working on this PR? If not, I'd be interested in continuing work on it.

**Comment by yaahc** [comment]
@Aaron1011 Right now this PR is on pause because we figured the next step would be to write an RFC based on the content of this WIP PR to hash out the final design before completing the PR.

@JDuchniewicz has started working on that RFC here: https://github.com/JDuchniewicz/rfcs/blob/backtrace-in-core/0000-backtrace-in-core.md

That said, I'd love to have any help we can get pushing this PR into a more complete state, I'm just worried that it may be for naught if we end up rejecting this design.

**Comment by bjorn3** [comment]
From the RFC:

> only that it is moved to core and if std is not linked, automatic backtrace handlers will be generated

How will these be generated? https://github.com/rust-lang/rust/pull/86844 tries to remove the allocator shim allowing `--emit obj` objects that link to liballoc to be linked directly without having rustc generate the allocator shim. This is currently already working for libcore, but a backtrace handler shim would now start requiring a shim to even be able to link to libcore, which is an impossible to avoid dependency. There are `#![no_std]` projects that don't use liballoc and link the object files themself that would break because of this as rustc doesn't get the chance to generate any shim.

**Comment by yaahc** [comment]
> From the RFC:
> 
> > only that it is moved to core and if std is not linked, automatic backtrace handlers will be generated
> 
> How will these be generated?

Not sure. This is one of the areas I was worried about and figured might tank this design. I don't really have a solution and was hoping people with more familiarity with linking would be able to help me figure out if it was viable.

**Comment by bjorn3** [comment]
Weak linkage would be the standard answer to implement this in different languages I think. There are some targets supported by rustc that don't support weak linkage though I believe.

**Comment by yaahc** [comment]
Yea, that's the same thing I ran into when trying weak linkage early in the process. 

I think I need to get a better idea of how this approach could cause issues for `no_std` projects. The current impl is closer to the `#[panic_handler]` attribute than the allocator shim as far as I can tell, but as is it still breaks no_std projects because they would need to define the hooks when previously they didn't, and I don't know if there's a `panic=abort` equivalent we could add for backtrace to indicate that the application never intends to capture backtraces. Either way @dtolnay is gonna look into getting an engineer working on this full time so we will have some help digging through the trade offs soon I hope.

**Comment by yaahc** [comment]
Closing this as no longer necessary. May re-open in the future if we end up deciding to keep `fn backtrace` on the Error trait.

**Comment by dvc94ch** [comment]
@yaahc can you summarize the decision for people not intimately familiar with all the discussions? Is rust not going to get support for backtraces outside of nightly?

**Comment by taqtiqa-mark** [comment]
Hmmm, the first I believe I could/would guess. I don't believe I would ever guess the second.... Likely contributing to increasing the Rust reputation as difficult or at least unintuitive?
My 2c.

**Comment by yaahc** [comment]
> Hmmm, the first example I believe I could/would guess. I don't believe I would ever guess the second.... Likely contributing to increasing the Rust-lang reputation as difficult or at least unintuitive? My 2c.

I appreciate the feedback, and yea, this is something I've been worried about. I'll keep trying to think about what I can do to make it easier to learn / understand / intuit, but I'm not really sure what I can do other than focus on documentation and learning resources. I think there may also possibly be a more intuitive / descriptive name than `context_ref`/`context_value` but so far I've struggled to come up with anything that is significantly more compelling...

**Comment by Alxandr** [comment]
Not sure if this has been discussed (I didn't see it here) - but might it be a good idea to provide a `backtrace` method on `&dyn Error` that calls this context-method under the hood? As this is probably going to be the most used property on `Error`, and would provide some ease of discovery?

**Comment by bjorn3** [comment]
The problem is that if `Error` is moved to libcore, `Backtrace` would then need to be too. However backtraces require an unwinder and some way tk ask the os for the unwind tables, which isn't possible with libcore.

**Comment by tarcieri** [comment]
Perhaps `Backtrace` could have a constructor method which extracts from a `dyn Error` context, e.g. `Backtrace::from_error`?

**Comment by dvc94ch** [comment]
I don't think this has to be overanalyzed. The main goal is that people can finally debug their applications without compiling with nightly and then realizing that some dependency did not capture a backtrace, then having to clone/patch the crate and unwrap all the ?'s.

It doesn't have to be beginner friendly to access the Backtrace as all beginners want to do is {} or {:?}

**Comment by tarcieri** [comment]
I think the most important things are to have backtrace-aware exception loggers, and as an added bonus where possible error reporting service hooks (e.g. Airbrake, Datadog, LogRocket, Sentry, Rollbar) which can accept a structured backtrace as part of an error report.

Ideally one should not be handrolling code to spelunk through a `dyn Error` to extract the backtrace. There should be a simple and obvious way to print the error along with its associated context, which may or may not include a backtrace.

**Comment by taqtiqa-mark** [comment]
> I'll keep trying to think about what I can do to make it easier to learn / understand / intuit ... and focus on teaching the one that is more powerful. It's a difficult balance :/

Agreed. "There are no solutions, only trade-offs"  

Here perhaps the doc's can motivate the novice by presenting an example exhibiting the power?  Showing they have head room, and trust they'll be able to recognize the forthcoming "Productivity" or "Performance" the example demonstrates.

If a new user is looking at Rust they likely have performance on their priority list (performance, reliability, productivity), and so they may be open to accepting the tradeoff if they are a little less productive to start with.

**Comment by dvc94ch** [comment]
I also don't think rust is that hard for beginners, even though it has that rap. You can clone your way out of lifetime issues, ownership is quite an intuitive concept, and the type system is built in a way that you can learn over time. Contrasting it with Haskell for example where if everything is built with monads and transformers and applicative monoids. Either you know what all those things are or you'll have a hard time with reading Haskell code.

**Comment by yaahc** [comment]
> Not sure if this has been discussed (I didn't see it here) - but might it be a good idea to provide a `backtrace` method on `&dyn Error` that calls this context-method under the hood? As this is probably going to be the most used property on `Error`, and would provide some ease of discovery?

This is a great idea! Though I think there are currently a couple of technical hurdles in the way. We already have something like this for `downcast` where we have to use a lang item to let the downcast methods live in `alloc` that include `Box` in their API. The only problem is `Backtrace` would not be able to live in `alloc` either, and I don't believe the compiler currently supports using lang items for three impls. The current interface only takes two impl IDs. Though as far as I know there's no reason that we couldn't extend this logic to allow three different impls. Alternatively we could do this as part of the proposed merger of `std`, `alloc`, and `core` under `--build-std` which would supersede the need for lang items.

**Comment by bjorn3** [comment]
> I don't believe the compiler currently supports using lang items for splitting an inherent impl across three crates.

AFAIK rustc does allow it. The float types are primitives and thus not defined in any existing crate. They have inherent impls in both libcore and liballoc.

**Comment by yaahc** [comment]
> > I don't believe the compiler currently supports using lang items for splitting an inherent impl across three crates.
> 
> AFAIK rustc does allow it. The float types are primitives and thus not defined in any existing crate. They have inherent impls in both libcore and liballoc.

That is consistent with my understanding but the issue is we need libcore, liballoc, and libstd, not just libcore and liballoc. I tried double checking the impl and it looks like it is still only two impls to me, though surprisingly the second one appears to be in libstd not liballoc, but either way I don't understand how rustc would allow this.

from compiler/rustc_hir/src/lang_items.rs:

```rust
    F32,                     sym::f32,                 f32_impl,                   Target::Impl,           GenericRequirement::None;
    F64,                     sym::f64,                 f64_impl,                   Target::Impl,           GenericRequirement::None;
    F32Runtime,              sym::f32_runtime,         f32_runtime_impl,           Target::Impl,           GenericRequirement::None;
    F64Runtime,              sym::f64_runtime,         f64_runtime_impl,           Target::Impl,           GenericRequirement::None;
```

from compiler/rustc_typeck/src/coherence/inherent_impls.rs:

```rust
            ty::Float(ty::FloatTy::F32) => {
                self.check_primitive_impl(
                    item.def_id,
                    lang_items.f32_impl(),
                    lang_items.f32_runtime_impl(),
                    "f32",
                    "f32",
                    item.span,
                    assoc_items,
                );
            }
```

from library/std/src/f32.rs:

```rust
#[cfg(not(test))]
#[lang = "f32_runtime"]
impl f32 {
```

from library/core/src/num/f32.rs
```rust
#[lang = "f32"]
#[cfg(not(test))]
impl f32 {
```

**Comment by bjorn3** [comment]
You can add an extra paramater to `check_primitive_impl` I think.

**Comment by yaahc** [comment]
> You can add an extra paramater to `check_primitive_impl` I think.

Ah, yea that was my original plan until @lcnr said he and the compiler team had a better plan that would supercede these lang items entirely. If that ends up running into issues though adding a parameter to the method will definitely be the approach I end up taking here.

**Comment by lcnr** [comment]
opened https://github.com/rust-lang/compiler-team/issues/487 and will work on implementing this over the next weeks

**Comment by taqtiqa-mark** [comment]
> this is something I've been worried about ...
> ...
> ... I think there may also possibly be a more intuitive / descriptive name than context_ref/context_value but so far I've struggled to come up with anything that is significantly more compelling

I used to think the 'Only two things in CS are hard...' quip was a pointed jab. Until I had to do it :)

Caveat: GitHub doesn't show any code for functions `context_*` so the following is speculative.

To my mind `context_*` is going to be encompassing (back, forward, sideways etc), `backtrace` is unidirectional.  

If the docs say that `context_*` can provide you with representation(s) of past-state and current-state (and/or future-state) then the idea **is** intuitive - `backtrace` is a special instance of that.  

If not, and `context_*` only relates to past-state, then my expectations about `context_*` are out of alignment with what it does.  
In case that is the situation: When I look at lists of synonyms (YMV) for "context" and "backtrace" the only synonym for "context" that intuitively (to my mind) connects to "backtrace" is background.  Not sure if `background_*` makes sense in the the context (sorry) of what these `context_*` functions are intended to do going forward:

```rust
fn extract_backtrace(error: &dyn Error) -> Option<&Backtrace> {
    error.background_ref::<Backtrace>()
}
```

**Comment by yaahc** [comment]
@taqtiqa-mark ah, yea, `context` isn't meant to be a synonym to backtrace or directly related to it. Instead it's supposed to be a generalization, where you can think of backtrace as a single kind of context on an error that you might care about. Context in this case meaning "information relevant to an error that is not itself part of the error message". Backtrace is context in so far as it acts like a map describing where in your codebase the error originated from, but the `.context_*` methods are meant to support more types than just backtrace. For example we want to be able to use it to grab things like `ExitCode`s from errors in `std::process::Termination` or things like suggestions and notes which could be displayed after the set of error messages in an error report, such as in this example from docs.rs/color-eyre:

```
color-eyre on  hooked [$!] is 📦 v0.5.0 via 🦀 v1.44.0
❯ cargo run --example custom_section
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/examples/custom_section`
Error:
   0: Unable to read config
   1: cmd exited with non-zero status code

Stderr:
   cat: fake_file: No such file or directory

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ SPANTRACE ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

   0: custom_section::output2 with self="cat" "fake_file"
      at examples/custom_section.rs:14
   1: custom_section::read_file with path="fake_file"
      at examples/custom_section.rs:58
   2: custom_section::read_config
      at examples/custom_section.rs:63

Suggestion: try using a file that exists next time
```
